### ISSUE TITLE
TF Lite Ondevice Training - Training Problem With Fashion MNIST

### ISSUE THREAD

**Issue by DonghakPark** [tensorflow/tensorflow#56251]
### Issue Type

Performance

### Source

source

### Tensorflow Version

tf 2.8

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 20.04

### Python version

3.8.10

### Bazel version

4.2.1

### GCC/Compiler version

9.4.0

### CUDA/cuDNN version

-

### GPU model and memory

-

### Current Behaviour?

```shell
refer site : https://www.tensorflow.org/lite/examples/on_device_training/overview

I'm Converting model to Tensorflow Lite and Testing on Linux Ubuntu 20.04. Trying Re-Training with signature('train'), Fashion MNIST

I remade the model several times, checked the shape of the data, but it was the same as in Python. However, proceed with the training, I can see the "nan" value, incorrect data as shown in the log below. What's the problem?

Here is my Build Command
.sh
bazel build --config=monolithic --config=noaws --config=nogcp --config=nohdfs --config=nonccl  --fat_apk_cpu=x86_64 --experimental_ui_max_stdouterr_bytes=1073741819 -c opt --cxxopt=--std=c++14 //tensorflow/lite:libtensorflowlite.so
```
ADD below to tensorflow/lite/BUILD
```.txt
"//tensorflow/lite/kernels:builtin_ops",
        "//tensorflow/lite/delegates/flex:delegate",
        "//tensorflow/lite/delegates/flex:exported_symbols.lds",
        "//tensorflow/lite/delegates/flex:version_script.lds",
```
```


### Standalone code to reproduce the issue

```shell
#include <cstdio>
#include <string>
#include <cstddef>
#include <cstdint>
#include <iostream>
#include <fstream>
#include <stdio.h>
#include <iomanip>
#include <getopt.h>
#include <ctime>
#include <typeinfo>
#include <algorithm>

#include "tensorflow/lite/interpreter.h"
#include "tensorflow/lite/interpreter_builder.h"
#include "tensorflow/lite/kernels/register.h"
#include "tensorflow/lite/model.h"
#include "tensorflow/lite/optional_debug_tools.h"
#include "tensorflow/lite/signature_runner.h"

#include "mnist/mnist_reader.hpp"

#define BATCH_SIZE 100
#define EPOCH 100

inline bool is_little_indian() {
	int val = 1;
	return *reinterpret_cast<char *>(&val) != 0;
}

uint32_t reverse_indian(uint32_t val) {
	val = ((val << 8) & 0xFF00FF00) | ((val >> 8) & 0xFF00FF);
	return (val << 16) | (val >> 16);
}

int main(int argc, char* argv[]) {

	const char* model_filename = argv[1];
    const int num_epochs = EPOCH;
    
    mnist::MNIST_dataset<std::vector, std::vector<uint8_t>, uint8_t> dataset =
        mnist::read_dataset<std::vector, std::vector, uint8_t, uint8_t>(MNIST_FASHION_DATA_LOCATION);

	//Loading the tflite retrainable model
	std::unique_ptr<tflite::FlatBufferModel> model = tflite::FlatBufferModel::BuildFromFile(model_filename);
	
	tflite::ops::builtin::BuiltinOpResolver resolver;
	tflite::InterpreterBuilder builder(*model, resolver);

	//Building the interpreter
	std::unique_ptr<tflite::Interpreter> interpreter;
	builder(&interpreter);

	//Getting the list of signatures
	auto signature_defs = interpreter->signature_keys();
		
    tflite::SignatureRunner* train_runner = interpreter->GetSignatureRunner("train");
    
    //Getting the input and output tensors names
    const std::vector<const char*>& input_names = train_runner->input_names();
    const std::vector<const char*>& output_names = train_runner->output_names();

    TfLiteTensor* train_input_x = train_runner->input_tensor((std::string(input_names[0])).c_str()); // x
    TfLiteTensor* train_input_y = train_runner->input_tensor((std::string(input_names[1])).c_str()); // y

    //Allocating tensors
    interpreter->AllocateTensors();
    train_runner->AllocateTensors();

    const int num_pixels = 28 * 28;
    const int num_batches = 60000 / BATCH_SIZE;

    float* input_x = train_input_x->data.f;
    float* input_y = train_input_y->data.f;

    std::cout << "[DEBUG][METADATA] Number of batches: " << num_batches << std::endl;
    std::cout << "[DEBUG][TRAINING] Training launched ... " << std::endl;
    
    
    // tflite::PrintInterpreterState(interpreter.get());
    for (int epoch = 0; epoch < num_epochs; ++epoch)
    {
        auto image_it = dataset.training_images.begin();
        auto label_it = dataset.training_labels.begin();
        float current_loss = 0;
        std::cout << "[DEBUG][TRAINING] Current epoch: " << epoch << std::endl;

        // float *fpixels{ new float[num_pixels * BATCH_SIZE]{}};
        // float labels[10 * BATCH_SIZE] = {0.0};
        float *fpixels{ new float[num_pixels]{}};
        float labels[10] = {0.0};

        for (uint32_t batch_index = 0; batch_index < num_batches; ++batch_index)
        {
            for (int index = 0; index < BATCH_SIZE; ++index)
            {   
                auto it = image_it->begin();

                // labels[int(*label_it) + (10 * index)] = 1.0;
                labels[int(*label_it)] = 1.0;
                for (int i=0; i<num_pixels; ++i)
                {   
                    // fpixels[i + (num_pixels * index)] = float(int(*it)/255.0);
                    fpixels[i] = float(float(*it)/255.0);
                    it++;
                }
                
                label_it++;
                image_it++;
                
                input_x = fpixels;
                input_y = labels;
                train_runner->Invoke();
                
            }
        }
        delete[] fpixels;
        const TfLiteTensor* output_tensor = train_runner->output_tensor((std::string(output_names[0])).c_str());
        float* output = output_tensor->data.f;
        current_loss = *(output);
        std::cout << "[DEBUG][TRAINING] Current Loss: " << current_loss << std::endl;
        
    }
    std::cout << "[DEBUG][TRAINING] Training finished successfully." << std::endl;
	
	return 0;
}
```

### Relevant log output

```shell
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
[DEBUG][METADATA] Number of batches: 600
[DEBUG][TRAINING] Training launched ... 
[DEBUG][TRAINING] Current epoch: 0
[DEBUG][TRAINING] Current Loss: -3.96266e+22
[DEBUG][TRAINING] Current epoch: 1
[DEBUG][TRAINING] Current Loss: -3.96266e+22
[DEBUG][TRAINING] Current epoch: 2
[DEBUG][TRAINING] Current Loss: -3.96266e+22
[DEBUG][TRAINING] Current epoch: 3
```

**Comment by AIML-ankit** [comment]
what was the approach can u share the code?

**Comment by DonghakPark** [comment]
> what was the approach can u share the code?

Unfortunately, I do not currently possess any code.
but as i remember you can call input tensor from tflite file and feed sequencially then they work correctly
